[TensorExpr] Add core classes for representing expressions and statements.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33220 [TensorExpr] Add IR Printer.
* #33219 [TensorExpr] Add IR visitor, IR mutator, and IR evaluator.
* **#33218 [TensorExpr] Add core classes for representing expressions and statements.**
* #33217 [TensorExpr] Add a class for representing data type.
* #33216 [TensorExpr] Add classes for memory management in tensor expressions.



Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

Commit:
[TensorExpr] Add core classes for representing expressions and statements.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add core classes for representing expressions and statements."


Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add core classes for representing expressions and statements."


Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add core classes for representing expressions and statements."


Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add core classes for representing expressions and statements."


Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

[ghstack-poisoned]

ZolotukhinM commented on test/cpp/tensorexpr/test_expr.cpp:
```diff
@@ -0,0 +1,58 @@
+namespace jit {
+using namespace torch::jit::tensorexpr;
+
+void testExprVectorAdd01() {
```
Most of the tests use IR evaluator, which is added in the subsequent commit, so the tests are also mostly added there.

Commit:
Update on "[TensorExpr] Add core classes for representing expressions and statements."


Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add core classes for representing expressions and statements."


Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add core classes for representing expressions and statements."


Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add core classes for representing expressions and statements."


Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add core classes for representing expressions and statements."


Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add core classes for representing expressions and statements."


Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add core classes for representing expressions and statements."


Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

[ghstack-poisoned]

dzhulgakov reviewed (approved):
What would be really helpful for onboarding on this code is to have large header comments or a single README file describing the core classes. As discussed offline, it can happen in a separate PR.

Commit:
Rebase on "[TensorExpr] Add core classes for representing expressions and statements."


Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

[ghstack-poisoned]

Commit:
Rebase on "[TensorExpr] Add core classes for representing expressions and statements."


Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add core classes for representing expressions and statements."




Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

[ghstack-poisoned]

Commit:
Rebase. on "[TensorExpr] Add core classes for representing expressions and statements."




Differential Revision: [D19848378](https://our.internmc.facebook.com/intern/diff/D19848378)

[ghstack-poisoned]
